pkg/controller/master/nodedrain/nodedrain_controller.go

```diff
@@ -293,22 +320,80 @@ func vmContainsCDRomOrContainerDisk(vmi *kubevirtv1.VirtualMachineInstance) bool
+	for _, vmi := range vmiList {
+		for _, condition := range vmi.Status.Conditions {
+			if condition.Type == kubevirtv1.VirtualMachineInstanceIsMigratable && condition.Status == corev1.ConditionFalse {
+				result := nonMigratableVM[condition.Reason]
+				result = append(result, namespacedVMName(vmi))
+				nonMigratableVM[condition.Reason] = result
+			}
+		}
 	}
-	labelSelector := labels.SelectorFromSet(labelsMap)
+	return nonMigratableVM
+}
 
-	vmiList, err := ndc.virtualMachineInstanceCache.List(corev1.NamespaceAll, labelSelector)
+func namespacedVMName(vmi *kubevirtv1.VirtualMachineInstance) string {
+	return fmt.Sprintf("%s/%s", vmi.Namespace, vmi.Name)
+}
+
+func splitNamespacedName(namespacedName string) (string, string) {
+	vmDetails := strings.Split(namespacedName, "/")
+	return vmDetails[0], vmDetails[1]
+}
+
+// CheckVMISchedulingRequirements checks if the VMI can be scheduled on another node
+// the function will check additional nodes that
+// * are able to satisfy the NodeSelectors terms specified in RequiredDuringSchedulingIgnoredDuringExecution
+// * and node is ready
+func (ndc *ControllerHandler) CheckVMISchedulingRequirements(originalNode *corev1.Node, vmiList []*kubevirtv1.VirtualMachineInstance) ([]string, error) {
+	var impactedVMS []string
+	nodeList, err := ndc.nodeCache.List(labels.Everything())
 	if err != nil {
-		return nil, fmt.Errorf("error listing VMI: %v", err)
+		return nil, fmt.Errorf("error listing nodes from nodeCache: %v", err)
 	}
-
-	var impactedVMI []string
 	for _, vmi := range vmiList {
-		if len(vmi.Spec.Domain.Devices.HostDevices) != 0 {
-			impactedVMI = append(impactedVMI, fmt.Sprintf("%s/%s", vmi.Namespace, vmi.Name))
+		var possibleNodes []string
+		if vmi.Spec.Affinity != nil && vmi.Spec.Affinity.NodeAffinity != nil && vmi.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+			nodeAffinitySelector, err := nodeaffinity.NewNodeSelector(vmi.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution)
+			if err != nil {
+				return nil, fmt.Errorf("error generating nodeAffinitySelector from node scheduling requirements: %v", err)
+			}
+			// identify if nodeAffinity can be met by other nodes and node is ready
+			for _, v := range nodeList {
+				if nodeAffinitySelector.Match(v) && v.Name != originalNode.Name && isNodeReady(v) {
+					possibleNodes = append(possibleNodes, v.Name)
+				}
+			}
+			// no valid node found that could meet the requirements
+			if len(possibleNodes) == 0 {
+				impactedVMS = append(impactedVMS, namespacedVMName(vmi))
+			}
 		}
 	}
-	return impactedVMI, nil
+	return impactedVMS, nil
+}
+
+func isNodeReady(node *corev1.Node) bool {
+	if node.Spec.Unschedulable {
+		return false
+	}
+
+	for _, v := range node.Status.Conditions {
+		if v.Type == corev1.NodeReady && v.Status == corev1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func findUniqueVMS(vms map[string][]string) []string {
```

Comment:
When I first saw the name of this function, I thought "wait, how are the VMs not already unique?" then I realised we're being passed a map of conditions to VM lists, because of course it's possible for there to be several conditions which would prevent a given VM from being migratable. Anyway, the point is, the name of the function kinda threw me :-) so I'd suggest maybe adding a comment about what's happening here, or, change the name (uhmmm... `getVMSfromConditionMap`?)